src/basic/virt.c

```diff
@@ -503,6 +505,14 @@ int detect_container(void) {
         } else if (r != -ENOENT)
                 return r;
 
+        /* The device number of an unnamespaced /proc/1/ns/{ns} always seems to be 3 or 4, while in a namespace it’s a
+         * higher number, apparently related to the number of namespaces. Let's use this as the last resort. */
+        if (lstat("/proc/1/ns/pid", &st) == 0)
```

Comment:
hmm, this sad limitation of this check is that it requires privileges... :-(

Also, it doesn't work for me...

On my 4.15 kernel I get this:
```
# stat /proc/1/ns/pid
  File: /proc/1/ns/pid -> pid:[4026531836]
  Size: 0         	Blocks: 0          IO Block: 1024   symbolic link
Device: 4h/4d	Inode: 6445506     Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:system_r:init_t:s0
Access: 2018-02-16 12:48:48.948488217 +0100
Modify: 2018-02-16 12:48:48.947488217 +0100
Change: 2018-02-16 12:48:48.947488217 +0100
 Birth: -
```

Not sure what's going on there...

Reply:
Why does it require privileges? I can stat that file as a regular user on my system without problems.

Also, I think the reason it doesn’t work for you could be that the patch is accidentally using the inode number instead of the device number?

Minimal test program derived from this patch:
```c
#include <stdio.h>
#include <sys/stat.h>

void main(int argc, char *argv[]) {
  struct stat st;
  if (lstat("/proc/1/ns/pid", &st) == 0)
    printf("%s: %svirtualization, st_dev is %d\n", argv[0], st.st_dev > 4 ? "" : "no ", st.st_dev);
}
```